yktrace/src/hwt/mapper.rs

```diff
@@ -129,15 +129,24 @@ impl HWTMapper {
                 }
             } else {
                 for irblock in irblocks.into_iter() {
-                    // A basic block may map to >1 *machine* basic block. If we see the same block
-                    // repeated, then we need to check if we re-executing the same basic block or
-                    // if we are executing next machine basic block for that same block. In the
-                    // latter case, we mustn't record the block in the trace again.
-                    if !(ret_irblocks.last() == Some(&irblock)
-                        && prev_block.as_ref().map(|x| x.first_instr())
-                            != Some(block.first_instr()))
-                    {
-                        ret_irblocks.push(irblock);
+                    if let Some(irblock) = irblock {
+                        // A basic block may map to >1 *machine* basic block. If we see the same
+                        // block repeated, then we need to check if we re-executing the same basic
+                        // block or if we are executing next machine basic block for that same
+                        // block. In the latter case, we mustn't record the block in the trace
+                        // again.
+                        if !(ret_irblocks.last() == Some(&irblock)
+                            && prev_block.as_ref().map(|x| x.first_instr())
+                                != Some(block.first_instr()))
+                        {
+                            ret_irblocks.push(irblock);
+                        }
+                    } else {
+                        // Part of a PT block mapped to a machine block in the LLVM block address
+                        // map, but the machine block has no corresponding IR blocks. This happens
+                        // because LLVM can introduce extra `MachineBasicBlock`s to help with
+                        // laying out machine code. For our purposes these extra blocks can be
+                        // ignored.
```

Comment:
The PR description is a bit vague: do we actually see this happen? If we haven't yet seen it happen, but *think* it can happen, I suggest putting `errx` or similar here so that we're notified of it and can check that we do the right thing if/when we see it happen.

Reply:
It does indeed happen when JITing ykcbf.

Reply:
How do we know that we can ignore these? Are they always fall-through or ... ?

Reply:
I can add a not in the commit message at squash time if you like.

Reply:
How does that relate to my questions?

Reply:
(we raced, will answer after lunch)

Reply:
> How do we know that we can ignore these?

My hypothesis is that if the LLVM IR captures all required program semantics, then extra blocks are either semantically ineffectual, or contain semantics transferred from other IR blocks. In both cases, the LLVM IR would remain stand alone.

> Are they always fall-through or ... ?

I think the block in question contained an `add`, which would suggest that the block is semantically meaningful, but I can't say if that was due to code moving out of a normal IR block and into one of these "extra" blocks or not. I'd have to dig deeper.

e83176b should make it clear that more work is required. It references an issue I've just created.